datafusion/sql/src/parser.rs

```diff
@@ -679,7 +679,25 @@ impl<'a> DFParser<'a> {
                     Keyword::PARTITIONED => {
                         self.parser.expect_keyword(Keyword::BY)?;
                         ensure_not_set(&builder.table_partition_cols, "PARTITIONED BY")?;
-                        builder.table_partition_cols = Some(self.parse_partitions()?);
+                        let peeked = self.parser.peek_nth_token(2);
+                        if peeked == Token::Comma || peeked == Token::RParen {
```

Comment:
Considering replacing this `if` with a more robust function that tries to apply a parsing rule and fallsback (undo consumed tokens) when rule fails.

Reply:
Yeah, it is not immediately clear why this logic works. I think at a minimum we should add a comment explaining the reasoning of this condition.

Reply:
I'm not super familiar with sqlparser crate, but  I don't think it allows rewinding tokens, we will have to implement a parsing rule that only uses peeks, which sounds really unnecessary. Will add a comment for now and maybe can find a better way.

Reply:
I think you can use the poorly named `consume_tokens` API:

https://docs.rs/sqlparser/latest/sqlparser/parser/struct.Parser.html#method.consume_tokens

Reply:
That looks promising. I'll open a follow up PR if it works out. Thanks for pointing it out.